### PR TITLE
Add two-hop WireGuard for iOS

### PR DESCRIPTION
**Pull request by pronebird**
- Add two-hop WireGuard integration for iOS
- Rework OSTunProvider to match the system API available on iOS for configuring tunnel interface, routes, dns etc..
- Add custom runner for wireguard on iOS. I am hoping we can harmonize the overall implementation between desktop and mobile but I have no energy to do that right now as the scope is only growing.

### CODE REVIEW

File: nym-vpn-apple/NymMixnetTunnel/PacketTunnelProvider.swift
Comment (rokas-ambrazevicius) [inline nym-vpn-apple/NymMixnetTunnel/PacketTunnelProvider.swift:49]: https://github.com/nymtech/nym-vpn-client/blob/94ba6cda9598710f42c8f17935de9dbb7642dadd/nym-vpn-apple/Services/Sources/Services/ConnectionManager/ConnectionStorage.swift#L29

Changing this to `.random`

Comment (rokas-ambrazevicius) [reply nym-vpn-apple/NymMixnetTunnel/PacketTunnelProvider.swift:48]: or here https://github.com/nymtech/nym-vpn-client/blob/97db4f4eb056d62fc5e725acda6f0423d23f366c/nym-vpn-apple/Services/Sources/Services/TunnelMixnet/Config/MixnetConfig.swift#L67

Comment (pronebird) [reply nym-vpn-apple/NymMixnetTunnel/PacketTunnelProvider.swift]: Rokas have you already fixed that right?

Comment (rokas-ambrazevicius) [reply nym-vpn-apple/NymMixnetTunnel/PacketTunnelProvider.swift]: Should be working in any case after this was merged:
https://github.com/nymtech/nym-vpn-client/pull/885

**Review by zaneschepke — APPROVED** [review]

**Review by zaneschepke — APPROVED** [review]

**Review by zaneschepke — APPROVED** [review]

**Review by zaneschepke — APPROVED** [review]

**Comment by pronebird** [comment]
@octol @neacsu @zaneschepke @rokas-ambrazevicius please leave your review early next week, preferably on Monday because we gotta move on with this and @zaneschepke is very close to bringing support for wg on android too.

File: .github/workflows/ci-nym-vpn-core.yml
Comment (octol) [inline .github/workflows/ci-nym-vpn-core.yml:206]: General comment _not_ for this PR:
The platform stuff is really reaching a point where we should try to take a step back and rework it. Also I'm kinda think that this part here is the result of building both mac and ios during a single job while splitting them into separate seems likely more appropriate (like how android is separate from linux).

File: nym-vpn-apple/MixnetLibrary/Package.swift
Comment (octol) [inline nym-vpn-apple/MixnetLibrary/Package.swift:27]: This will break after the next nightly build?

File: nym-vpn-core/nym-vpn-lib/src/lib.rs
Comment (octol) [inline nym-vpn-core/nym-vpn-lib/src/lib.rs:55]: nitpick: separate config.rs file or similar?

File: nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs
Comment (octol) [inline nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs:8]: I like this, wrapping channels in a strong type. Imho we should do this more often

File: nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs
Comment (octol) [inline nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs:1]: Side comment: for unbounded receivers/senders there is no confusion, but for bounded senders/receivers then the type has the same name for mpsc/broadcast/oneshot etc so it's nice to not import the whole path so it's easier to see in the code what channel it is

File: nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs
Comment (octol) [inline nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs:76]: Unrelated side comment: we should implement `Default` for `MixnetClientConfig` in the future

File: nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs
Comment (octol) [inline nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs:133]: Side comment for the future:
Imho we should switch BandwidthController to use CancellationToken. Also, BandwidthController calls `mixnet_client.disconnect()` I just spotted, not sure what the implication of that is ...

File: nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs
Comment (octol) [inline nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs:127]: I have a strong suspicion that this is a no-op since we pass in an external `task_manager.subscribe()` when creating the mixnet client, and the way to disconnect the mixnet_client is to call `task_manager.signal_shutdown()`

File: nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs
Comment (octol) [inline nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs:26]: What's the thinking when deciding between placing stuff in this mod vs `platform`? Is `platform` something that will be phased out in time?

**Review by octol — APPROVED** [review]
Looking great! Well done!

I left some side comments for stuff I noticed while reading the code, some which refers to code not touched by this PR. Feel free to do what you want with them

File: nym-vpn-apple/MixnetLibrary/Package.swift
Comment (rokas-ambrazevicius) [reply nym-vpn-apple/MixnetLibrary/Package.swift:27]: This will break. I believe this is a temp fix to get the Testflight out the door.

Comment (octol) [reply nym-vpn-apple/MixnetLibrary/Package.swift:27]: If you need a custom build of the vpn lib that is non-temporary you can use a named tag like `nym-vpn-core-ios-testflight`. But maybe if we merge this on Monday we can create a new release core then and use that?

Comment (rokas-ambrazevicius) [reply nym-vpn-apple/MixnetLibrary/Package.swift:27]: That would be ideal 👍

File: .github/workflows/ci-nym-vpn-core.yml
Comment (pronebird) [reply .github/workflows/ci-nym-vpn-core.yml:206]: Yeah that's legit. Let's add a card for this and attack it in a separate PR.

File: nym-vpn-apple/MixnetLibrary/Package.swift
Comment (pronebird) [reply nym-vpn-apple/MixnetLibrary/Package.swift:27]: we'll fix this with a follow up pr

File: nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs
Comment (pronebird) [reply nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs:1]: Correct. Bounded are different story.

File: nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs
Comment (pronebird) [reply nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs:8]: Note that this struct implements `OSDefaultPathObserver` trait, making it a conduit for messages received via FFI boundary.

File: nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs
Comment (pronebird) [reply nym-vpn-core/nym-vpn-lib/src/mobile/ios/default_path_observer.rs:26]: Simply a logical grouping of related types. Everything under `mobile/ios` in essence is a platform-specific stuff. I didn't want to move things too apart since it breaks connection between types.

File: nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs
Comment (pronebird) [reply nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs:133]: Yeah what BandwidthController does is it seems that it just waits for task manager to shutdown before killing the mixnet client. I am not quite sure why it's done this way, this could be re-arranged.

File: nym-vpn-core/nym-vpn-lib/src/lib.rs
Comment (pronebird) [reply nym-vpn-core/nym-vpn-lib/src/lib.rs:55]: I'll do that in a separate PR to unblock people.

File: nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs
Comment (pronebird) [reply nym-vpn-core/nym-vpn-lib/src/mobile/runner.rs:127]: It might be, I ran out of time to investigate this any closer.

**Pull request merged**
